src/sage/rings/polynomial/polynomial_element.pyx

```diff
@@ -816,18 +830,23 @@ cdef class Polynomial(CommutativePolynomial):
                 try:
                     # Note that we may be calling a different implementation that
                     # is more permissive about its arguments than we are.
-                    cst = cst(*args, **kwds)
-                    eval_coeffs = True
+                    top = top(*args, **kwds)
                 except TypeError:
                     if args: # bwd compat: nonsense *keyword* arguments are okay
                         raise TypeError("Wrong number of arguments")
+                else:
+                    eval_coeffs = True
 
             # Evaluate the coefficients, then fall through to evaluate the
             # resulting univariate polynomial
 
             if eval_coeffs:
+                new_base = parent(top)
+                # this value for the common parent of the evaluated
```

Comment:
I don't understand what you mean here.

Reply:
Well, I am not 100% sure what the code is doing exactly. It seem to try to guess a new base ring, depending not only on the existing base ring, but on the substituted coefficients. But it does that by looking only at the substitution in one of the coefficients. Before my changes, it was looking at the substitution in the constant term, and fails if this was zero. I changed it to use the substitution of the leading term. This works a little better.

Of course, the only thing making complete sense would be to look at the common parent of all substituted coeffs. But this may have a large speed penalty..

Reply:
> But it does that by looking only at the substitution in one of the coefficients.

Yes, but for this whole business (of evaluating the coefficients of a polynomial and getting a new polynomial) to make sense, evaluation (of the coefficients) should be a ring morphism, so it should be enough to look at the evaluation of one coefficient, or in fact of any element of the base ring.

So as I said in the other comment, I think it is okay to work around the fact that some of the base rings we may want to work with have an evaluation function that does not respect this constraint. However, I'm still not sure what you mean by “this value [...] only works for monomials”.